reframe/frontend/executors/__init__.py

```diff
@@ -130,6 +131,37 @@ def __init__(self, case, listeners=[]):
         # Test case has finished, but has not been waited for yet
         self.zombie = False
 
+        # Timestamps
+        self._timestamp = {
+            'start_setup': None,
+            'finish_setup': None,
+            'start_compile': None,
+            'finish_compile': None,
+            'start_run': None,
+            'finish_run': None,
+            'start_sanity': None,
+            'finish_sanity': None,
+            'start_performance': None,
+            'finish_performance': None,
+            'start_cleanup': None,
+            'finish_cleanup': None
+        }
+
+    @property
+    def duration(self):
```

Comment:
I would also make that accept the phase.

Reply:
If I change the timestamps to be taken in the `_safe_call()` function, what would `duration('run')` return? `timestamp['finish_run'] - timestamp['start_run' ]` or `timestamps['finish_wait'] - timestamps['start_run' ]`?

Reply:
Because in the output run is not the same as the internal `run` phase of the pipeline.

Reply:
I think that's entirely up to us. You can return duration of the whole if the phase is for example `'run+wait'`. We can give it any name.

Reply:
The `'run+wait'` is maybe inconsistent because if it is what is called `run` now.
```
timestamp['run+wait'] != timestamp['run'] + timestamp['wait']
```
But we print it as `run` duration.

Reply:
Don't confuse what you report and how you get it. We can name it as we like when we are reporting it. I can see where the confusion is coming from, since the "run" phase from the [user perspective](https://reframe-hpc.readthedocs.io/en/latest/pipeline.html#the-run-phase) is the whole thing (run_spawn + run_wait) but inside the framework the `run` is essentially the run_spawn. Implementation-wise now, it's easier if the `phase` argument refers to what we do internally, i.e., the run_spawn subphase.

Reply:
Ok then, if you agree I will name:
```
duration['run'] = timestamp['finish_run'] - timestamp['start_run']
duration['wait'] = timestamp['finish_wait'] - timestamp['start_wait']
duration['run_pipeline'] = timestamp['finish_wait'] - timestamp['start_run']
duration['compile_pipeline'] = timestamp['finish_compile_wait'] - timestamp['start_compile']
```
`run_pipeline` and `compile_pipeline` could also be `'running'` and `'waiting'`

Reply:
I would rather call the combined phases as `run_complete` and `compile_complete`.

Reply:
Ok, changing them to this, thanks. There a couple more small issues.
1. What should the `total` be after these changes?
Before it was the sum of all the individual "phases". Should we also make this `timestamp['finish_last_stage'] - timestamp['start_setup']` or as it was before (summing only the phases we report)?
2. If I insert the timestamps in `_safe_call()` instead of each individual function and add directly an end timestamp based on the _failed_stage in fail(), then (in case the test fails in the run_spawn) `duration['run_complete']` should be:
`duration['run_complete'] = timestamp['finish_run'] - timestamp['start_run']` instead of `timestamp['finish_wait'] - timestamp['start_run']`

Reply:
1. The total should *not* be the sum. I wanted to comment that before. We lose the "other" in such a case. So it should be as you propose, `last_stage - setup_start`. But for last stage why not adding just a new `pipeline_end` timestamp, so that the total will always be `pipelilne_end - setup_start`? Can you also inverse the timestamp names? `setup_start` instead of `start_setup`?

2. Using my suggestion above, you can always pick the `pipeline_end` if you can't find the end-of-phase timestamp.